js/foam/dao/AuthenticatedDAO.js

```diff
@@ -0,0 +1,210 @@
+      // authenticator to massage it for this principal, and return it.
+      // The authenticator might return null, in which case we return an error.
+      var self = this;
+      var principal = opt_X && opt_X.principal;
+      if (!principal) {
+        sink && sink.error && sink.error('Illegal find');
+        return;
+      }
+
+      this.delegate.find(id, {
+        error: sink && sink.error && sink.error.bind(sink),
+        put: function(obj) {
+          self.authenticator.massageForRead(function(obj) {
+            if (obj === null) {
+              sink && sink.error && sink.error('Failed to find');
+            } else {
+              sink && sink.put && sink.put(obj);
+            }
+          }, principal, obj);
+        }
+      });
+    },
+    function select(sink, options, opt_X) {
+      // On select, we first decorate the delegate to avoid wasting effort.
+      // Then we run the select(), and massage each output value for reading
+      // before sending it to the sink. nulls are dropped, as though they don't
+      // exist. (That should be rare, unless the authenticator can't express all
+      // of its conditions as a where() condition).
+      // TODO(braden): Conditions from decorateForSelect() are fine, but if
+      // massageForRead() will return null for any retrieved objects even with
+      // the extra conditions from decorateForSelect, then the skip and limit
+      // counts will be thrown off. This shouldn't matter most of the time,
+      // since decorateForSelect will be comprehensive, and massageForRead will
+      // never return null on the selected objects.
+      if (!sink) sink = [].sink;
+      var future = afuture();
+      var self = this;
+      var principal = opt_X && opt_X.principal;
+      if (!principal) {
+        sink && sink.error && sink.error('Illegal select.');
+        future.set(sink);
+        return future.get;
+      }
+
+      aseq(
+        function(ret) {
+          self.authenticator.decorateForSelect(ret, principal, self.delegate);
+        },
+        function(ret, dao) {
+          dao.select({
+            error: sink && sink.error && sink.error.bind(sink),
+            put: function(obj) {
+              self.authenticator.massageForRead(function(massaged) {
+                if (massaged !== null) {
+                  sink && sink.put && sink.put(massaged);
+                }
+              }, principal, obj);
+            },
+            eof: function() {
+              sink && sink.eof && sink.eof();
+              ret();
+            }
+          }, options, opt_X);
+        }
+      )(function() {
+        future.set(sink);
+      });
+
+      return future.get;
+    },
+    function removeAll(sink, options, opt_X) {
+      // TODO(braden): Implement me. Use the select+remove scheme.
+    },
+    function where(query) {
+      return (this.Y || X).lookup('FilteredDAO_').create({ query: query, delegate: this });
```

Comment:
Pass this.Y as the second parameter to create.  This should be done in AbstractDAO as well but you can leave that till later.

I would also reference these as the globals that they are.

FilteredDAO.create({...}, this.Y);